lib/spack/spack/spec.py

```diff
@@ -3411,6 +3441,11 @@ def satisfies(self, other: Union[str, "Spec"], deps: bool = True) -> bool:
         lhs_edges: Dict[str, Set[DependencySpec]] = collections.defaultdict(set)
         mock_nodes_from_old_specfiles = set()
         for rhs_edge in other.traverse_edges(root=False, cover="edges"):
+            # Skip checking any conditional edge that is not satisfied
+            if rhs_edge.when != Spec() and not self.satisfies(rhs_edge.when):
```

Comment:
By just reading the code, it seems to me the very fundamental assumption on subset semantic:
```
abstract_spec.satisfies(x) is True -> concrete_spec.satisfies(x) is True 
```
doesn't hold anymore for conditional deps. A very simple reproducer:
```python
Spack version 1.0.0.dev0
Python 3.13.0, Linux x86_64
>>> from spack.spec import Spec
>>> s = Spec("hdf5")
>>> constrain = Spec("%[when='+mpi' virtuals=zlib-api] zlib")
>>> s.satisfies(constrain)
True
>>> s.concretize()
==> Warning: <console>:1: `Spec.concretize` is deprecated and will be removed in version 1.0.0. Use `spack.concretize.concretize_one` instead.
>>> s.satisfies(constrain)
False
```
Did we work out the consequences of this ?

Reply:
Just to add to this, conditional specs act as a filter on the set of solutions just like abstract specs do. As an example, suppose `foo` has a boolean variant `x` and a package `bar` exists without any variants.

The abstract spec `foo` induces the set of concrete specs:

```
{foo ~x %bar, foo +x %bar, foo ~x, foo +x}
```

The conditional spec

```
foo %[when=+x] bar
```

induces the following set of concrete specs

```
{foo ~x %bar, foo +x %bar, foo ~x}
```

Since that set is missing the concrete spec `foo +x`, the `foo` does **not** satisfy `foo %[when=+x] bar`.

The right-hand side is more constrained than the left-hand side.

It's important to get this right.

Reply:
To differentiate the two cases we probably need:
```
if not spec.intersects(when_condition):
   # fine to skip edge, it's the hdf5~mpi case
elif not spec.satisfies(<condition>):
   # This should be False
   return False
```